README.md
This is the Dilbert Strip Index (finder_dsi), used by the *Dilbert
Strip Finder* at <http://www.bfmartin.ca/finder>.

It contains:

- The source data in JSON format used to index the content of
  the Dilbert comic strips so a full-text search can be performed.

- Some programs written in Ruby to load the data into a MariaDB
  database, test the validity of the data, and some command-line
  utilities to report and maintain the data.

You can reach me at <http://www.bfmartin.ca/contact>

This project is dedicated to all fans of Dilbert and to Scott Adams in
particular for creating the comic strip.


# Copyright and Licenses

## JSON files

The JSON data files (dsistrips.json and dsibooks.json) are copyright
1999 - 2017 by Byron F. Martin. They are licensed under the [Creative
Commons Attribution 2.5 Canada
license](http://creativecommons.org/licenses/by/2.5/ca/).  The short
version is that you can do whatever you want with these files as long
as you give me credit for creating them. A link to my site will be
good, though it's not required.

## Programs written by me

The program files (Ruby and Rakefile files) that were written by me
(which is everything except the bundled module `lib/dsi/stem.rb`) are
donated to the public domain. Do whatever you want with them.

## Bundled module

A bundled Ruby module for word stemming (`lib/dsi/stem.rb`) is
copyright by its author(s). Please refer to its documentation for its
license information. It was obtained from
<http://tartarus.org/~martin/PorterStemmer/ruby.txt>.


# JSON files

## Dates

All dates are in the format YYYY-MM-DD.


## dsistrips.json

This is the main data file that describes each Dilbert strip.  For
each strip, the JSON file contains:

- Date of newspaper publication.

- Synopsis, a one or two sentence description of the strip.

- Subject (optional), the main topic.

- Keywords (optional), many words to describe the action and important
  concepts.

- Characters (optional), the characters with speaking parts.

- Saga (optional), to mark the beginning of three or more strips with
  a common theme.

- Notes (optional), to describe important items about the strip, like
  the first appearance of a character.

- Comment (optional), text about the strip for my own purposes and is
  not meant to be displayed in search results.


## dsidialog.json

This data is not included here but is generated from some text files
available on the Internet. The programs format the data for easy
searching.

This file is generated by running the command

        rake dialog:prepare

This will download the necessary file(s) from their respective sites
and generate dsidialog.json. See the Rakefile for details on how this
happens.


## dsibooks.json

NOTE: This file is retired, and is *not* maintained any more. You can
find all strips online, so finding them in books is superfluous. This
file stays, but has no books past 2014.

This describes each Dilbert comic collection and its contents.  For
each book, the file contains (all are required):

- a one word code id for the book.

- the book title

- the layout of a week of strips, beginning with Sunday, and
  separated by commas.  Most books are laid out as 1,3,3 which means:

  - Sunday on one page

  - three more strips on the next page

  - three more strips on the next page.

- A list containing the start and end dates and the page number that
  they start on. This is a list because sometimes strips are not in
  chronological order.

# Ruby Requirements

If you like to play around with the Ruby programming language, there
are some example programs included in this package. To make them work
you will need to have the following software packages installed. Your
operating system's package manager should be able to help with these.


- Ruby, at least version 2.3.

- Rake

- The following ruby gems

  - json        (to parse JSON files)
  - net/http    (to fetch dialog file)
  - optimist    (to parse command line options)
  - rake_notes  (shows TODOs and FIXMEs in code)

  Try: `gem install <gemname>`


# Database

Some programs refer to a database. The SQL will work with MariaDB, and
uses the following table definitions. The search feature relies on
MariaDB's full text indexing.

# bin directory

There are several programs included to work with the DSI data.

If the program name begins with 'dsi', then its purpose is to maintain
or report on the dsi data.

- dsi-generate-dialog.rb

  this will read raw dialog files (as downloaded from different web
  locations) and merge them into a standard format JSON file.

- dsi-key.rb

  this will read the dsistrips json file and format it for printing.
  it takes the keywords or characters or subjects, and prints a list
  of dates that contain that item.

- dsi-notes.rb

  this will read the dsistrips json file and print all items with
  notes.  Useful for browsing and debugging.


If the program name begins with 'finder', then it generates data to
update the Finder web site.

- finder-load.rb

  reads dsistrips and creates lines to be loaded into the bfmartin.ca
  database.  See the database schema in the README.  also loads dialog
  if available, otherwise inserts nulls.

- finder-reindex.sh

  a wrapper to do everything. it will download all required files,
  reformat them, and load the database.

# Feedback

If you have questions or feedback, you can reach me at
<http://www.bfmartin.ca/contact>

You can visit the *Dilbert Strip Finder* web site at
<http://www.bfmartin.ca/finder>


Changelog
Version 5.0.2 2018-11-07
	- change from trollop gem to optimist, its replacement.

Version 5.0.1 2017-11-07
	- renamed files finder_dsi to finderdsi to match the library name.

Version 5.0.0 2017-11-07
	- removed functionality from the programs for the books. Since all
	the strips are online, I don't think the book and page number are
	useful anymore. this means:
	** the dsibooks.json file will remain, but is unmaintained after 2014.
	** bin/finder-books-update.rb and bin/finder-books-page-update.rb
	are removed.
	** library methods for book and page lookup based on dates are
	removed.
	- instead of two dialogue files available online, only one still
	exists. stop trying to fetch the removed one.
	- removed method fetch_latest_strip_date. it didn't work out for
	me. this means bin/finder_load_incremental.rb is also removed as
	it depended on this method.
	- lots of style cleanups in the ruby code. the most impactful one
	is renaming Finder_DSI library to FinderDSI.
	- bin/dsi-key.rb command line option parsing changed from optparse
	to trollop gem.
	- bin/dsi-generate-dialog.rb no longer has an option to specify an
	output file. it always sends output to stdout, and can be
	redirected to a file on the command line. so removed dependency on
	optparse gem.
	- because of the removal of the book functionality, there are
	backward imcompatibilities and the major version number has been
	bumped.

Version 4.0.1 2012-11-01
	- data/dsistrip.json: added strips to 2016-10-31.

Version 4.0.0 2012-08-01
	- data/dsistrip.json: added strips to 2012-08-01.
	- created a gem called finder_dsi. The version number jumps to a
	major release because of this, and its format changes slightly to
	be compatible with gems. renamed some files to keep in line with
	the new name.
	- imported code into github. some files have been re-arranged to
	better suit their layout.
	- bin/*.rb: renamed program files to be more consistent.
	- README converted to markdown format. Some minor cleanups as well.
	- some refactoring as part of the gem creation. Some APIs have
	changed to be backwards incompatible. Since I don't know of anyone
	who uses these Ruby APIs, it probably doesn't matter much.

Version 3.05 2012-02-01
	- data/dsistrips.json: added strips to 2012-01-31.
	- data/dsibooks.json: added book "How's that Underling Thing
	Working Out for You?".
	- data/dsibooks.json: fixed date range for "I'm Tempted to Stop
	Acting Randomly". No one noticed as far as I know. Does anyone use
	the book / page reference anymore? The online archive is much more
	immediate

Version 3.04 2011-10-02
	- data/dsistrips.json: added strips to 2011-09-30.
	- data/dsistrips.json: found and replaced 4 more strips containing
	a synopsis composed of words selected from the dialog.
	- lib/dsi/entry.rb: new
	- lib/dsi.rb: added methods fetch_latest_strip_date and has_dialog?.
	fetch_latest_strip_date is experimental only for now
	- bin/key.rb: fixed bug where subjects weren't displayed properly.
	- bin/incremental.rb: new
	- lib/*.rb: added a subdirectory called 'dsi' and moved most
	library files to there. this matches other ruby module layouts.
	- bin/*.rb, lib/**/*.rb and test/*.rb: more refactoring based on ruby
	code quality checkers. Pretty much done as much as can be done.

Version 3.03 2011-07-31
	- data/dsistrips.json: added strips to 2011-07-31. Spelling and
	grammar corrections throughout.
	- data/dsibooks.json: added "I'm Tempted to Stop Acting Randomly".
	- data/dsibooks.json: changed id of "Random Acts of Management"
	from "random" to "acts" because of a new book called "I'm Tempted
	to Stop Acting Randomly".
	- bin/*rb, lib/*.rb and test/*.rb: refactoring based on ruby code
	quality checkers (reek, roodi, flay, flog and saikuro). still
	more to do, though.

Version 3.02 2010-07-31
	- data/dsistrips.json: added strips to 2010-07-31. Spelling and
	other minor corrections throughout.
	- data/dsibooks.json: added books "Freedom's Just Another Word for
	People Finding Out You're Useless" and "14 Years of Loyal Service
	in a Fabric-Covered Box".
	- bin/load.rb: added word stemming to the database load.
	- lib/dsi.rb: added method dsistriphash (and test in test-data.rb).
	- lib/stem.rb: added a porter-stemming module to this
	distribution. this excellent library was aquired from
	http://tartarus.org/~martin/PorterStemmer/
	- doc/create-mysql-tables.sql: add dsitopsearches create
	statement.

Version 3.01 2009-06-17
	- dsistrips.json: added strips to 2009-06-14
	- dsistrips.json: refinements, additions, spelling corrections to
	older DSI entries
	- key.rb: fixed bug that caused a no method error
	- key.rb: added options -a and -n

Version 3.00 2009-05-01
	- Combined dsibooks and dsistrips into a single download
	- Added strips to 2009-04-30, though the ones past 2007-02-02 are
	just random collections of important words in the dialog.
	- spelling and wording corrections in dsistrips.json
	- Converted XML files into JSON.
	- Added books "Don't Stand Where the Comet is Assumed to Strike
	Oil", "The Fluorescent Light Glistens Off Your Head", "Thriving
	on Vague Objectives", "Try Rebooting Yourself", "Positive
	Attitude", "This is the Part Where You Pretend to Add Value"
	- Removed java programs, added ruby ones
	- Changed licenses to the Creative Commons Attribution 2.5 Canada
	License for the dsi JSON files and public domain for programs

Version 2.03 2004-01-01
	- Added strips to 2003-12-31.
	- Added book "Words You Don't Want to Hear at Your Performance
	Review"

Version 2.02 2003-04-30
	- Added strips to 2003-04-30.
	- Added book "When Body Language Goes Bad"

Version 2.01 2003-01-01
	- Added strips to 2002-12-31.

Version 2.00 2002-09-01
	- Added strips to 2002-08-31.
	- Changed the XML DTD to a XML schema.
	- The XML date format changed from YYYYMMDD to YYYY-MM-DD+00:00
	to be compatible with the XML schema date format.
	- Book published year has changed format from YYYY to YYYY+00:00
	for the same reason.

Version 1.07 2002-04-03

lib/dsi/stem.rb
#! /local/ruby/bin/ruby
#
# $Id: stemmable.rb,v 1.2 2003/02/01 02:07:30 condit Exp $
#
# See example usage at the end of this file.
#

module Stemmable

  STEP_2_LIST = {
    'ational'=>'ate', 'tional'=>'tion', 'enci'=>'ence', 'anci'=>'ance',
    'izer'=>'ize', 'bli'=>'ble',
    'alli'=>'al', 'entli'=>'ent', 'eli'=>'e', 'ousli'=>'ous',
    'ization'=>'ize', 'ation'=>'ate',
    'ator'=>'ate', 'alism'=>'al', 'iveness'=>'ive', 'fulness'=>'ful',
    'ousness'=>'ous', 'aliti'=>'al',
    'iviti'=>'ive', 'biliti'=>'ble', 'logi'=>'log'
  }
  
  STEP_3_LIST = {
    'icate'=>'ic', 'ative'=>'', 'alize'=>'al', 'iciti'=>'ic',
    'ical'=>'ic', 'ful'=>'', 'ness'=>''
  }


  SUFFIX_1_REGEXP = /(
                    ational  |
                    tional   |
                    enci     |
                    anci     |
                    izer     |
                    bli      |
                    alli     |
                    entli    |
                    eli      |
                    ousli    |
                    ization  |
                    ation    |
                    ator     |
                    alism    |
                    iveness  |
                    fulness  |
                    ousness  |
                    aliti    |
                    iviti    |
                    biliti   |
                    logi)$/x


  SUFFIX_2_REGEXP = /(
                      al       |
                      ance     |
                      ence     |
                      er       |
                      ic       | 
                      able     |
                      ible     |
                      ant      |
                      ement    |
                      ment     |
                      ent      |
                      ou       |
                      ism      |
                      ate      |
                      iti      |
                      ous      |
                      ive      |
                      ize)$/x


  C = "[^aeiou]"         # consonant
  V = "[aeiouy]"         # vowel
  CC = "#{C}(?>[^aeiouy]*)"  # consonant sequence
  VV = "#{V}(?>[aeiou]*)"    # vowel sequence

  MGR0 = /^(#{CC})?#{VV}#{CC}/o                # [cc]vvcc... is m>0
  MEQ1 = /^(#{CC})?#{VV}#{CC}(#{VV})?$/o       # [cc]vvcc[vv] is m=1
  MGR1 = /^(#{CC})?#{VV}#{CC}#{VV}#{CC}/o      # [cc]vvccvvcc... is m>1
  VOWEL_IN_STEM   = /^(#{CC})?#{V}/o                      # vowel in stem

  #
  # Porter stemmer in Ruby.
  #
  # This is the Porter stemming algorithm, ported to Ruby from the
  # version coded up in Perl.  It's easy to follow against the rules
  # in the original paper in:
  #
  #   Porter, 1980, An algorithm for suffix stripping, Program, Vol. 14,
  #   no. 3, pp 130-137,
  #
  # See also http://www.tartarus.org/~martin/PorterStemmer
  #
  # Send comments to raypereda@hotmail.com
  #
  
  def stem_porter

    # make a copy of the given object and convert it to a string.
    w = self.dup.to_str
    
    return w if w.length < 3
    
    # now map initial y to Y so that the patterns never treat it as vowel
    w[0] = 'Y' if w[0] == ?y
    
    # Step 1a
    if w =~ /(ss|i)es$/
      w = $` + $1
    elsif w =~ /([^s])s$/ 
      w = $` + $1
    end

    # Step 1b
    if w =~ /eed$/
      w.chop! if $` =~ MGR0 
    elsif w =~ /(ed|ing)$/
      stem = $`
      if stem =~ VOWEL_IN_STEM 
        w = stem
	case w
        when /(at|bl|iz)$/             then w << "e"
        when /([^aeiouylsz])\1$/       then w.chop!
        when /^#{CC}#{V}[^aeiouwxy]$/o then w << "e"
        end
      end
    end

    if w =~ /y$/ 
      stem = $`
      w = stem + "i" if stem =~ VOWEL_IN_STEM 
    end

    # Step 2
    if w =~ SUFFIX_1_REGEXP
      stem = $`
      suffix = $1
      # print "stem= " + stem + "\n" + "suffix=" + suffix + "\n"
      if stem =~ MGR0
        w = stem + STEP_2_LIST[suffix]
      end
    end

    # Step 3
    if w =~ /(icate|ative|alize|iciti|ical|ful|ness)$/
      stem = $`
      suffix = $1
      if stem =~ MGR0
        w = stem + STEP_3_LIST[suffix]
      end
    end

    # Step 4
    if w =~ SUFFIX_2_REGEXP
      stem = $`
      if stem =~ MGR1
        w = stem
      end
    elsif w =~ /(s|t)(ion)$/
      stem = $` + $1
      if stem =~ MGR1
        w = stem
      end
    end

    #  Step 5
    if w =~ /e$/ 
      stem = $`
      if (stem =~ MGR1) ||
          (stem =~ MEQ1 && stem !~ /^#{CC}#{V}[^aeiouwxy]$/o)
        w = stem
      end
    end

    if w =~ /ll$/ && w =~ MGR1
      w.chop!
    end

    # and turn initial Y back to y
    w[0] = 'y' if w[0] == ?Y

    w
  end


  #
  # make the stem_porter the default stem method, just in case we
  # feel like having multiple stemmers available later.
  #
  alias stem stem_porter

end



#
# Make this script executable, and send it words on stdin, one per
# line, and it will output the stemmed versions to stdout.
#
if $0 == __FILE__ then
  class String
    include Stemmable
  end

  # the String class, and any subclasses of it you might have, now know
  # how to stem things.

  $stdin.each do |word|
    puts word.stem
  end
end

